subsys/logging/log_core.c

```diff
@@ -995,31 +996,37 @@ void z_vrfy_z_log_string_from_user(uint32_t src_level_val, const char *str)
 		return;
 	}
 
-	/*
-	 * Validate and make a copy of the source string. Because we need
-	 * the log subsystem to eventually free it, we're going to use
-	 * log_strdup().
-	 */
-	len = z_user_string_nlen(str, (level == LOG_LEVEL_INTERNAL_RAW_STRING) ?
-				 CONFIG_LOG_PRINTK_MAX_STRING_LENGTH :
-				 CONFIG_LOG_STRDUP_MAX_STRING, &err);
+	is_imm = IS_ENABLED(CONFIG_LOG_IMMEDIATE);
+	is_raw = !is_imm && IS_ENABLED(CONFIG_LOG_PRINTK) && (level == LOG_LEVEL_INTERNAL_RAW_STRING);
+	is_nrm = !is_imm && !is_raw;
+	max_len = (level == LOG_LEVEL_INTERNAL_RAW_STRING) ?
+						CONFIG_LOG_PRINTK_MAX_STRING_LENGTH :
+						CONFIG_LOG_STRDUP_MAX_STRING;
 
-	Z_OOPS(Z_SYSCALL_VERIFY_MSG(err == 0, "invalid string passed in"));
-	Z_OOPS(Z_SYSCALL_MEMORY_READ(str, len));
+	if (is_nrm) {
+		char kstr[MAX(CONFIG_LOG_PRINTK_MAX_STRING_LENGTH, CONFIG_LOG_STRDUP_MAX_STRING) + 1];
 
-	if (IS_ENABLED(CONFIG_LOG_IMMEDIATE)) {
-		log_string_sync(src_level_union.structure, "%s", str);
-	} else if (IS_ENABLED(CONFIG_LOG_PRINTK) &&
-		   (level == LOG_LEVEL_INTERNAL_RAW_STRING)) {
-		struct log_msg *msg;
+		err = z_user_string_copy(kstr, (char *)str, max_len);
+		Z_OOPS(Z_SYSCALL_VERIFY_MSG(err == 0, "invalid string passed in"));
+		
+		str = log_strdup(kstr);
```

Comment:
"Why don't I just do `"%.*s"` instead?"

Because log_strdup itself is the main issue here...  I could make a version of it that takes an explicit max length (`log_strdup_s`?) I suppose.